source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -131,8 +131,19 @@ void RedisCluster::DnsDiscoveryResolveTarget::startResolve() {
       [this](std::list<Network::Address::InstanceConstSharedPtr>&& address_list) -> void {
         active_query_ = nullptr;
         ENVOY_LOG(trace, "async DNS resolution complete for {}", dns_address_);
-        parent_.redis_discovery_session_.registerDiscoveryAddress(address_list, port_);
-        parent_.redis_discovery_session_.startResolve();
+        if (address_list.empty()) {
+          parent_.info_->stats().update_empty_.inc();
+          if (!resolve_timer_) {
```

Comment:
Hmm, yikes, I'm not sure how I caught this before but you are resolving DNS in a tight loop in working case, which is very bad. You need to use the timer in all cases. Please rework the code.

Reply:
the redis_discovery_session is different from the "DNS discovery".  The DNS is only used in the initial startup that seed the list of addresses.  After that the DNS is not used at all.  In the redis_discovery_session, a timer is always used.

Reply:
I'm not sure I follow. In the below success case you start the resolve again. Doesn't that create a loop?

Reply:
In the success case, the start resolve is on a different class.  In that class, it call "cluster slots" on a random redis node periodically.  It doesn't call DNS resolver again.  The DNS resolver is only used to seed the list when the cluster is started.

Reply:
I see, OK, thanks for the explanation. Please add more comments and also rename the method names to be more clear about what is going on in each stage. Thanks.

/wait